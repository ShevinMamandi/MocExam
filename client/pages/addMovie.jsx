import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import { fetchJSON } from "../http";
import { logout } from "../logout";

export function AddMovie() {
  const [title, setTitle] = useState("");
  const [plot, setPlot] = useState("");
  const [year, setYear] = useState("");
  const navigate = useNavigate();

  async function handleSubmit(e) {
    e.preventDefault();
    await fetchJSON("/api/movies", {
      method: "post",
      json: { title, year, plot },
    });
    setTitle("");
    setYear("");
    setPlot("");
    navigate("/movies/list");
  }

  return (
    <form onSubmit={handleSubmit}>
      <div>
        <button onClick={logout}>Logout</button>
      </div>
      <h1>Add Movie</h1>
      <div>
        Title:
        <input value={title} onChange={(e) => setTitle(e.target.value)} />
      </div>
      <div>
        Year:
        <input value={year} onChange={(e) => setYear(e.target.value)} />
      </div>
      <div>Plot:</div>
      <div>
        <textarea value={plot} onChange={(e) => setPlot(e.target.value)} />
      </div>
      <button>Submit</button>
    </form>
  );
}
