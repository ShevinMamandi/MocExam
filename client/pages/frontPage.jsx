import { Link } from "react-router-dom";
import React from "react";

export function FrontPage() {
  return (
    <div>
      <h1>Login</h1>
      <ul>
        <li>
          <Link to="/login">Login</Link>
        </li>
      </ul>
    </div>
  );
}