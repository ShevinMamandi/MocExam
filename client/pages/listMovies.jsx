import { useLoader } from "../useLoader";
import { fetchJSONData } from "../http";
import { logout } from "../logout";
import React from "react";

export function ListMovies() {
  const { loading, error, data } = useLoader(async () => {
    return fetchJSONData("/api/movies");
  });

  if (loading) {
    return <div>Loading.... </div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div>{error.toString()}</div>
      </div>
    );
  }

  return (
    <div>
      <div>
        <button onClick={logout}>Logout</button>
      </div>
      <h1>List movies</h1>
      {data.map((movie) => (
        <MovieCard key={movie.title} movie={movie} />
      ))}
    </div>
  );
}
function MovieCard({ movie: { title, plot, year } }) {
  return (
    <>
      <h3>{title}</h3>
      <h4>{year}</h4>
      <div>{plot}</div>
    </>
  );
}
