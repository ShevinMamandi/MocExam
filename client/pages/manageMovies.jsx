import { useLoader } from "../useLoader";
import { fetchJSONData } from "../http";
import { logout } from "../logout";
import { Link } from "react-router-dom";
import React from "react";

export function ManageMovies() {
  const { loading, data, error } = useLoader(async () => {
    return await fetchJSONData("/api/login");
  });
  if (loading) {
    return <div>Please wait...</div>;
  }
  if (error) {
    return <div>Error! {error.toString()}</div>;
  }

  return (
    <div>
      <div>
        <button onClick={logout}>Logout</button>
      </div>
      <div>Profile for: {data.name}</div>
      <br />
      <h1>Movie Database</h1>
      <ul>
        <li>
          <Link to={"/movies/list"}>List movies</Link>
        </li>
        <li>
          <Link to={"/movies/new"}>Add new movie</Link>
        </li>
      </ul>
    </div>
  );
}
