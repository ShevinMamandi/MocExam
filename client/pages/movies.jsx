import { Route, Routes } from "react-router-dom";
import { ListMovies } from "./listMovies";
import { AddMovie } from "./addMovie";
import React from "react";
import { fetchJSONData } from "../http";

export function Movies() {
  async function listMovies() {
    return await fetchJSONData("/api/movies");
  }
  return (
    <Routes>
      <Route path={"/list"} element={<ListMovies listMovies={listMovies} />} />
      <Route path={"/new"} element={<AddMovie />} />
    </Routes>
  );
}
