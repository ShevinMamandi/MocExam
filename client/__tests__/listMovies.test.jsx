import { ListMovies } from "../pages/listMovies";
import ReactDOM from "react-dom";
import React from "react";
import { act } from "react-dom/test-utils";

describe("ListMovies component", () => {
  it("show loading screen", () => {
    const domElemenet = document.createElement("div");
    ReactDOM.render(<ListMovies />, domElemenet);
    expect(domElemenet).toMatchSnapshot();
  });

  it("show movies", async () => {
    const movies = [{ title: "movie 1" }, { title: "movie 2" }];
    const domElemenet = document.createElement("div");
    await act(async () => {
      ReactDOM.render(<ListMovies listMovies={() => movies} />, domElemenet);
    });
    expect(
      Array.from(domElemenet.querySelectorAll("h3")).map((e) => e.innerHTML)
    ).toEqual(["movie 1", "movie 2"]);
    expect(domElemenet).toMatchSnapshot();
  });

  it("show error message", async () => {
    const movies = [{ title: "movie 1" }, { title: "movie 2" }];
    const domElemenet = document.createElement("div");
    await act(async () => {
      ReactDOM.render(
        <ListMovies
          listMovies={() => {
            throw new Error("Something went wrong");
          }}
        />,
        domElemenet
      );
    });
    expect(domElemenet.querySelector("#error-text").innerHTML).toEqual(
      "Error: Something went wrong"
    );
    expect(domElemenet).toMatchSnapshot();
  });
});
