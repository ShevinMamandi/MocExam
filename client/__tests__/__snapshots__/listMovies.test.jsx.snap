// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListMovies component show error message 1`] = `
<div>
  <div>
    <h1>
      Error
    </h1>
    <div
      id="error-text"
    >
      Error: Something went wrong
    </div>
  </div>
</div>
`;

exports[`ListMovies component show loading screen 1`] = `
<div>
  <div>
    Loading.... 
  </div>
</div>
`;

exports[`ListMovies component show movies 1`] = `
<div>
  <div>
    <div>
      <button>
        Logout
      </button>
    </div>
    <h1>
      List movies
    </h1>
    <h3>
      movie 1
    </h3>
    <h4 />
    <div />
    <h3>
      movie 2
    </h3>
    <h4 />
    <div />
  </div>
</div>
`;
